File: design-pattern-js/chainOfResponsibility/Support.js
class Support {
    name
    next
    constructor(name) {
        this.name = name;
    }

    setNext(next) {
        this.next = next;
        return next;
    }

    support(trouble) {
        if (resolve(trouble)) {
            done(trouble);
        } else if (this.next) {
            next(trouble);
        } else {
            fail(trouble);
        }
    }

    resolve(trouble) { }

    done(trouble) {
        console.log(`${trouble.name} is resolved by ${this.name}`)
    }

    fail(trouble) {
        console.log(`${trouble.name} cannot be resolved.`)
    }
}

File: design_pattern-py/support.py
class Support:
    def __init__(self, name):
        self.name = name

    def setNext(self, next):
        self.next = next

    def support(self, trouble):
        if self.resolve(trouble):
            self.done(trouble)
        elif self.next != None:
            self.next.support(trouble)
        else:
            self.fail(trouble)

    def resolve(self, trouble):
        pass

    def done(self, trouble):
        print(f'{trouble.name} is resolved by {self.name}.')

    def fail(self, trouble):
        print(f'{trouble.name} cannot be resolved.')
